runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+	if err != nil {
+		return err
+	}
+
+	err = releaseConn()
+	if err != nil {
+		return err
+	}
+
+	// replicate
+	err = d.syncBackup(ctx, tableName)
+	if err != nil {
+		return err
+	}
+	d.writeDirty = false
+	// remove old version
+	_ = d.deleteBackup(ctx, tableName, version)
+	return d.syncRead(ctx)
+}
+
+// AlterTableColumn implements drivers.OLAPStore.
+func (d *db) AlterTableColumn(ctx context.Context, tableName, columnName, newType string) error {
+	d.logger.Debug("AlterTableColumn", slog.String("table", tableName), slog.String("column", columnName), slog.String("typ", newType))
+	d.writeMu.Lock()
+	defer d.writeMu.Unlock()
+	conn, release, err := d.acquireWriteConn(ctx)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = release()
+	}()
+
+	return d.alterTableColumn(ctx, conn, release, tableName, columnName, newType)
+}
+
+func (d *db) alterTableColumn(ctx context.Context, conn *sqlx.Conn, releaseConn func() error, tableName, columnName, newType string) error {
+	version, exist, err := tableVersion(d.writePath, tableName)
+	if err != nil {
+		return err
+	}
+
+	if !exist {
+		return fmt.Errorf("table %q does not exist", tableName)
+	}
+
+	d.writeDirty = true
+	_, err = conn.ExecContext(ctx, fmt.Sprintf("ALTER TABLE %s.default ALTER %s TYPE %s", safeSQLName(dbName(tableName)), safeSQLName(columnName), newType))
+	if err != nil {
+		return err
+	}
+
+	// rename to new version
+	newVersion := fmt.Sprint(time.Now().UnixMilli())
+	err = os.Rename(filepath.Join(d.writePath, tableName, version), filepath.Join(d.writePath, tableName, newVersion))
+	if err != nil {
+		return err
+	}
+
+	// update version.txt
+	err = os.WriteFile(filepath.Join(d.writePath, tableName, "version.txt"), []byte(newVersion), fs.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	err = releaseConn()
+	if err != nil {
+		return err
+	}
+
+	// replicate
+	err = d.syncBackup(ctx, tableName)
+	if err != nil {
+		return err
+	}
+	d.writeDirty = false
+	// remove old version
+	_ = d.deleteBackup(ctx, tableName, version)
+	return d.syncRead(ctx)
+}
+
+func (d *db) syncRead(ctx context.Context) error {
+	entries, err := os.ReadDir(d.writePath)
+	if err != nil {
+		return err
+	}
+
+	tableVersion := make(map[string]string)
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+
+		// Check if there is already a table with the same version
+		writeVersion, exist, _ := d.writeTableVersion(entry.Name())
+		if !exist {
+			continue
+		}
+		tableVersion[entry.Name()] = writeVersion
+		readVersion, _, _ := d.readTableVersion(entry.Name())
+		if writeVersion == readVersion {
+			continue
+		}
+
+		d.logger.Debug("Sync: copying table", slog.String("table", entry.Name()))
+		err = copyDir(filepath.Join(d.readPath, entry.Name()), filepath.Join(d.writePath, entry.Name()))
+		if err != nil {
+			return err
+		}
+	}
+
+	handle, err := d.openDBAndAttach(ctx, true)
+	if err != nil {
+		return err
+	}
+
+	var oldDBHandle *sqlx.DB
+	d.readMu.Lock()
```

Comment:
As I understand it, this might depend on all current reads draining first? It's not a good idea to have `Lock` calls for mutexes that may take a long time to become available because the code stops responding to `ctx` cancellation in this case.

Reply:
Yeah I thought about draining reads but could not find a satisfactory solution to avoid that. 
For context cancellations the only solution that comes to my mind is calling below function after CRUD operations to update read handle.
```
func (d *db) reopen(ctx context.Context, deletedTable string) error {
	done := make(chan struct{})
	go func() {
		d.writeMu.Lock()
		defer d.writeMu.Unlock()

		// work is done
		defer close(done)

		// chhose a timeout so that goroutine does not get stuck
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()
		handle, err := d.openDBAndAttach(ctx, true, deletedTable)
		if err != nil {
			d.logger.Error("reopen: unable to open db", slog.String("error", err.Error()))
		}

		// take lock on reads and swap the read handle
		var oldDBHandle *sqlx.DB
		d.readMu.Lock()
		oldDBHandle = d.readHandle
		d.readHandle = handle
		d.readMu.Unlock()

		// close old read handle
		if oldDBHandle != nil {
			err = oldDBHandle.Close()
			if err != nil {
				d.logger.Warn("error in closing old read handle", slog.String("error", err.Error()))
			}
		}

		// drop deleted table
		if deletedTable != "" {
			_ = d.deleteLocalTable(deletedTable, "")
		}
	}()

	select {
	case <-done:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}
```

Reply:
Yeah I also thought about reads draining first but could not think of any way to solve this. If we starve the `reopen` in favour of any new read query then there is a chance that in a heavily used dashboard data never refresh.

Regarding the issue of not responding to the cancellation I think we can borrow the `reopen` logic in current duckdb driver in case of fatal errors i.e. mark db as `reopen` and return from write. 
In this case the reads may still get blocked but writes will return early. This should be okay because writes getting blocked can block the reconciler. This also has an advantage that `reopen` can not get failed due to context cancellations. What do you think ?

Reply:
NVM. I think I found a satisfactory solution looking at `runtime/pkg/ctxsync` :)

Reply:
> I also thought about reads draining first but could not think of any way to solve this

With the previously discussed logic of having the write-side be responsible for creating new versions and the read-side responsible for cleaning up old versions on local, then can you avoid needing a long-running mutex at all?

For example, the read-side could track the number of current queries per table version. When the count for a version goes to 0, it checks if the table version is no longer the current one, and if it isn't, then it deletes the file. 

This would enable write queries to run concurrently with read queries with "snapshot isolation", i.e. where read queries resolve based on a snapshot of the database at the time when the query started.

Reply:
I realize this may be problematic when using a single read handle for all reads (to ensure correct resource limits). I can think of two workarounds for that, what do you think of these:
1. Instead of creating views of the attached tables in the read handle, create them as temporary views in each open connection on the read handle (since temporary views are connection-specific). This enables different concurrent queries to reference different versions of a table while still being subject to the same resource limits.
2. Alternatively, create a separate schema inside the read handle for each "generation" of the database (and bump the "generation" each time there's a new write; this would just be an in-memory counter that is reset on each restart). This enables a query to "use" a specific generation and not worry if a new generation is created with an updated table. New queries would always use the latest generation.

I think I like option 2 best. I think this would enable concurrent reads/writes without the need to ever drain connections before starting a new read. I think that would be nice, and maybe it's simple enough, but not sure if it is overthinking it. What do you think?

Reply:
> For example, the read-side could track the number of current queries per table version. When the count for a version goes to 0, it checks if the table version is no longer the current one, and if it isn't, then it deletes the file.

May be I am wrong(or overthinking) but I think this has the problem that in a heavily used dashboard data never refresh. With this we can not guarantee that queries will start looking at new data in acceptable amount of time in all cases.

> I think I like option 2 best. I think this would enable concurrent reads/writes without the need to ever drain connections before starting a new read. I think that would be nice, and maybe it's simple enough, but not sure if it is overthinking it. What do you think?

Yeah I think this is an excellent solution. Though I think we will need to take care of two things:
1. We may need to detach all dbs else we may leak memory if we just drop older schema. I can evaluate. It is fine today because the handle is closed which takes care of freeing memory as well.
2. Do garbage collection of old tables only when no queries read data from older "generation".

Reply:
> I think this has the problem that in a heavily used dashboard data never refresh

So actually as soon as there's a new version, the next query that starts should use the new version. But any query that is already running can keep referencing the old version. So in fact this will lead to new data being served as soon as possible, instead of waiting for old queries to drain.

The tricky thing is for a new query to access the new version while old queries still access the old version at the same time. This is the problem my options 1 and 2 above tries to solve. Does that make sense?

> 1. We may need to detach all dbs else we may leak memory if we just drop older schema. I can evaluate. It is fine today because the handle is closed which takes care of freeing memory as well.

Aren't DBs attached globally, not per schema? So wouldn't it be sufficient to detach the old versions before dropping the schema for the old generation?

> 2. Do garbage collection of old tables only when no queries read data from older "generation".

Yes this would be the essential step, and I think can be solved with a simple map and the aforementioned proposal of:

>  track the number of current queries per table version. When the count for a version goes to 0, it checks if the table version is no longer the current one, and if it isn't, then it deletes the file

Reply:
> Does that make sense?

Yes that makes sense. 

> Aren't DBs attached globally, not per schema? So wouldn't it be sufficient to detach the old versions before dropping the schema for the old generation?

Yes just outlining things that need to be taken care of. :)